mesonbuild/compilers/mixins/clike.py

```diff
@@ -688,7 +688,7 @@ def get_define(self, dname: str, prefix: str, env: 'Environment',
         # Get the preprocessed value after the delimiter,
         # minus the extra newline at the end and
         # merge string literals.
-        return self._concatenate_string_literals(p.stdout.split(delim + '\n')[-1][:-1]), cached
+        return self._concatenate_string_literals(p.stdout.split(delim + '\n')[-1][:-1]).strip(), cached
```

Comment:
I'm not sure we should blindly `strip()` the output, what if we had `#define FOO "   foo    "` ? I think a fix like the one proposed in https://github.com/mesonbuild/meson/pull/10182 is safer.

Reply:
Actually that was a bad example because output includes `"` chars of course. Maybe this is a good, as @eli-schwartz said in https://github.com/mesonbuild/meson/pull/10182#issuecomment-1080640115, we don't currently have a way to distinguish between undefined and defined with empty value, which is a separate issue.

LGTM then?

Reply:
Here's a cute little rework of that code: 5105fa556485197110104cc5230a310ef91d646b

It produces stuff like this:
```
Code:
 
        #include <concat.h>
        #ifndef TEST_CONCAT_6
        # define TEST_CONCAT_6
        #endif
        MESON_GET_DEFINE_DELIMITER(TEST_CONCAT_6)MESON_GET_DEFINE_DELIMITER
-----------
Command line: `c++ '-I/home/eschwartz/git/meson/test cases/common/132 get define/.' /tmp/tmp5073sxqu/testfile.cpp -E -P -D_FILE_OFFSET_BITS=64 -P -O0 -fpermissive` -> 0
stdout:
MESON_GET_DEFINE_DELIMITER("ab\" \"cd")MESON_GET_DEFINE_DELIMITER
-----------
the value extracted from p.stdout='        MESON_GET_DEFINE_DELIMITER("ab\\" \\"cd")MESON_GET_DEFINE_DELIMITER\n' with delim='MESON_GET_DEFINE_DELIMITER' delimlen=27 and start=27 is ('"ab\\" \\"cd"', False)
Fetching value of define "TEST_CONCAT_6" : "ab\" \"cd" 
```

I don't think double-guarding on both sides is really necessary. The main thing I want to draw attention to is that, consider the following code:
```
#define FOO    barify   )

        #ifndef FOO
        # define FOO
        #endif
MESON_GET_DEFINE_DELIMITER(FOO)MESON_GET_DEFINE_DELIMITER
```

`gcc -E -P testfile.c`

```
MESON_GET_DEFINE_DELIMITER(barify ))MESON_GET_DEFINE_DELIMITER
```

If you define FOO as:
```
#define FOO    barify   
```

note trailing whitespace

the stdout is
```
MESON_GET_DEFINE_DELIMITER(barify)MESON_GET_DEFINE_DELIMITER
```

define it as:
```
#define FOO "   barify   "      "lel"
```

output:
```
MESON_GET_DEFINE_DELIMITER("   barify   " "lel")MESON_GET_DEFINE_DELIMITER
```

As far as the C preprocessor is concerned, whitespace is not part of the define, it is not part of a preprocessing token at all. Whitespace distinguishes between separate tokens, and semantically collapses down to nothing. If you define a macro to have the value of two tokens separated by a whole bunch of whitespace, it collapses it down to two tokens, then inserts *one* space during formatting of the output.

This makes sense, really, because a C/C++ token cannot have trailing whitespace, and as you noted in your second comment, the presence of quote characters means the token ends in a quote character and internally contains whitespace as a *string value*.

This is a long-winded way of saying that I believe stripping is semantics-preserving here.

Reply:
> Here's a cute little rework of that code: [5105fa5](https://github.com/mesonbuild/meson/commit/5105fa556485197110104cc5230a310ef91d646b)

Are parenthesis needed? `{delim}{dname}{delim}` seems the simplest and safest approach here.

Reply:
Oh no, you definitely do not want no separation between delim and dname -- the current version in git master uses a newline, you could switch it for a single blank space but IMO that's no advantage at all.

If you write out the source file as:
```
MESON_GET_DEFINE_DELIMITERFOOMESON_GET_DEFINE_DELIMITER
```

Then you will get back:
```
MESON_GET_DEFINE_DELIMITERFOOMESON_GET_DEFINE_DELIMITER
```

Because it is not a macro and cannot be replaced.

...

The commit message from my experiment mentions:
> which is handy since this also lets us distinguish the open and the close from each other.

Reply:
Oops you're right indeed.